### ISSUE TITLE
Fedora link stales

### ISSUE THREAD

**Issue by lvroyce** [kimchi-project/kimchi#605]
Many fedora mirrors (such as mirrors.kernel.org and fedora.mirrors.tds.net)deleted fedora 18 iso. So in distro.d/fedora.json, we cannot use it any more, I'm wondering if deleting these items is a good choice.

**Comment by cd1** [comment]
Issues of this type have already been filed (e.g. #301) and others will be filed again whenever we need to remove old distros and add new ones. Besides that, it's not very useful to still keep a template for Fedora 18 when there already are Fedora 19, Fedora 20, Fedora 21 (and soon Fedora 22)...

I propose we query the newest distro versions online and show them when the user wants to create a remote template. I have this small script which demonstrates this idea:

```python
#!/usr/bin/python

import os.path
from ftplib import FTP

host = 'ftp.osuosl.org'

ftp = None

try:
    ftp = FTP(host)
    ftp.login()
    releases = ftp.nlst('pub/fedora/linux/releases')
    
    for r in filter(lambda x: os.path.basename(x) != 'test', releases):
        # Fedora >= 21 has a new directory structure
        if int(os.path.basename(r)) >= 21:
            isos = ftp.nlst('%s/Workstation/x86_64/iso/' % r)
        else:
            isos = ftp.nlst('%s/Fedora/x86_64/iso/' % r)

        for i in filter(lambda x: x.endswith('.iso'), isos):
            print 'ftp://%s/%s' % (host, i)
finally:
    if ftp is not None:
        ftp.quit()
```

Running it on my laptop prints the following output:

```
vianac@kitkat ~ $ time python fedora.py
ftp://ftp.osuosl.org/pub/fedora/linux/releases/20/Fedora/x86_64/iso/Fedora-20-x86_64-DVD.iso
ftp://ftp.osuosl.org/pub/fedora/linux/releases/20/Fedora/x86_64/iso/Fedora-20-x86_64-netinst.iso
ftp://ftp.osuosl.org/pub/fedora/linux/releases/21/Workstation/x86_64/iso/Fedora-Live-Workstation-x86_64-21-5.iso

real	0m6.274s
user	0m0.043s
sys	0m0.017s
```

This is just a sample. We would need to add more details to it if we want that idea on Kimchi (e.g. add more servers / mirror list, filter out the 'netinst' ISO), but we can see that it works well. Using something like that, we would have a distro available to use as a remote template _as soon as it becomes available online_. And we could keep only the latest n versions and remove the oldest ones automatically. IMO, this would be a great (and distinguishing) feature.

We might need to update the algorithm to fetch the info from the server once in a while (e.g. when a distro changes its distribution paths, like Fedora did after version 21) but that would be much less frequent than updating the distros URLs manually as we do now, especially considering the number of remote distros we support - and the number of more distros we _could_ support if we do this automatically. Opening the remote template creation dialog would also take more time. As you can see, the script took ~6 seconds to run here. If we start a thread to query this for every distro we display, I guess we would still take that same time approximately (and _not_ that amount of time for each distro). It's slower than what we have now, but I think it's worth it because we're getting content online and the user will have to wait even more to install the remote ISO anyway. The regular user, who doesn't care/know about remote templates, won't be affected by this because they won't even open the remote template creation dialog.

What do you think?

**Comment by laggarcia** [comment]
The idea looks OK to me. I'd rather not want to wait 6 seconds to have this information, but using some kind of cache might not work very well also.

However, you cannot look at only the x86_64 versions of the distro. You need something more generic, at least.

**Comment by cd1** [comment]
Sure, those features can be done in the final code; the script above is just to make sure the idea is valid and possible for implementation. The final code should be more robust.

**Comment by lvroyce** [comment]
I think we can run @cd1's code in build phase, so that .json files will be filled with valid links in packages. And 6s is not a long time for build.
But we do need some kind of hard code because the directory structure of ftp may change. We can only avoid hard code by writing a piece of code like "ftp deep scan".

**Comment by lvroyce** [comment]
And in kimchi server filtering out stale links to guarantee users won't create vm with invalid iso streaming.

**Comment by cd1** [comment]
An "FTP deep scan" doesn't sound like a good approach because it would take _a lot_ more time as, usually, an FTP server holds thousands of files for many different purposes. And that's why I think we won't have many problems with directory renames on the FTP mirrors because that's the purpose of mirrors, they host content in an expected layout, suitable for automation. Well, that's how I expect it to work :sweat_smile: 

And as we'd get live content from the FTP server, we wouldn't have stale links at all. When an ISO file gets removed from the server (e.g. Fedora 18), we won't fetch that file anymore and it just won't be listed as a remote ISO.

**Comment by cd1** [comment]
The issue regarding the Fedora stale links (which was causing test errors) has been fixed by 13de741f140ad7f1a239c882160443de6da8ec23. The conversation regarding the enhancement on the  current approach should continue on issue #611.
